pkg/policy/policy.go

```diff
@@ -117,13 +130,50 @@ func (p *Policies) PoliciesByKind(kind string) (map[string]string, error) {
 	return policies, nil
 }
 
-// TODO: use loaded
+func (p *Policies) getCachedHash(kind string) (string, error) {
+	// without cache, we don't keep the hashes
+	if p.cache == nil {
+		return "", nil
+	}
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+	value, err := p.cache.Get(kind)
+	if err != nil {
+		if !errors.Is(err, gcache.KeyNotFoundError) {
+			return "", fmt.Errorf("failed to get hash from cache: %v", err)
+		}
+		return "", nil
+	}
+	if hash, ok := value.(string); ok {
+		return hash, nil
+	}
+	return "", nil
+}
+func (p *Policies) setCachedHash(kind, hash string) error {
+	if p.cache == nil {
+		return nil
+	}
+	return p.cache.SetWithExpire(kind, hash, p.cacheExpire)
+}
```

Comment:
We're accessing the policy cache here, we will need a lock or else we risk a race condition between this and `getCachedHash()`

Reply:
Speaking of which, we should add a test for exercising this policy cache and run that test with the `-race` go test flag.

Reply:
Actually I realized we have some tests in pkg/policy/policy_test.go, but I'm unsure why we didn't see the issue with sorting over there? Is it because we don't assert the hash?

Reply:
> We're accessing the policy cache here, we will need a lock or else we risk a race condition between this and `getCachedHash()`

no, there is a lock under the hood `cache.SetWithExpire`.

Reply:
> I'm unsure why we didn't see the issue with sorting over there? Is it because we don't assert the hash?

I think we have a really small policies in the test.

Reply:
> > We're accessing the policy cache here, we will need a lock or else we risk a race condition between this and `getCachedHash()`
> 
> no, there is a lock under the hood `cache.SetWithExpire`.

But that lock is different from the mutex we are acquiring on the policies structure isn't it?

If gcache is inherently goroutine safe as it claims, why do we need https://github.com/aquasecurity/trivy-operator/blob/7d0e0e41332489e2bb56640086af9d1f54058592/pkg/policy/policy.go#L138

Reply:
> > I'm unsure why we didn't see the issue with sorting over there? Is it because we don't assert the hash?
> 
> I think we have a really small policies in the test.

I would assume as long as we have more than one, the order should become indeterministic  and the test should fail.

Reply:
Maybe we can add the tests I added in the other PR here? https://github.com/aquasecurity/trivy-operator/pull/2520/files#diff-b1b1d0d360676f39ad9629e53e7227915be6509110b1e38f3e8711cd69ca0686R49-R95

Reply:
> If gcache is inherently goroutine safe as it claims, why do we need

it seems you're right. we can remove this RLock and `mutex`, because `getValue` already uses a lock:
https://github.com/bluele/gcache/blob/d8b7e051c564c174fea6ef60d180abf601099015/lru.go#L124

Reply:
> I would assume as long as we have more than one, the order should become indeterministic and the test should fail.

last week i tried to add policies one-by-one, and if them becomes more 10, a map gives different result.
my suggestion is Go re-orders hash maps. I can deep it if needs

Reply:
> If gcache is inherently goroutine safe as it claims, why do we need

done https://github.com/aquasecurity/trivy-operator/pull/2526/commits/478086d8d14b50b553aa92d44225200848b4b152

Reply:
> Maybe we can add the tests I added in the other PR here?

oh, sure. I just thought we'll merge #2520 too. I meat we'll keep both PRs.